### PR TITLE
Kotlin implemtentation of minicap

### PR DESCRIPTION
**Pull request by pcrepieux**
This is an attempt to write a minicap version in kotlin and
use higher level api that should hopefully make it easier
to support many os version (and the future ones).
It also is likely to prevent the reccurent "Vector<> have
different types" errors.
The implementation is not complete and a few changes will have
to be done on stf.

### CODE REVIEW

File: experimental/app/src/main/java/io/devicefarmer/minicap/utils/DisplayManagerGlobal.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import android.annotation.SuppressLint
+import android.util.Size
+
+/**
+ * Get DisplayInfo through DisplayManagerGlobal singleton (similar to what is done in MinitouchAgent)
+ */
+@SuppressLint("PrivateApi")
+object DisplayManagerGlobal {
+    private var displayManager: Any?
+    private var CLASS: Class<*>?
```
Comment (koral--) [inline experimental/app/src/main/java/io/devicefarmer/minicap/utils/DisplayManagerGlobal.kt:27]: It's rather named like `clazz`

File: experimental/build.gradle [hunk 2]
```diff
@@ -0,0 +1,29 @@
+// Top-level build file where you can add configuration options common to all sub-projects/modules.
+buildscript {
+    ext.kotlin_version = "1.4.31"
+    repositories {
+        google()
+        jcenter()
```
Comment (koral--) [inline experimental/build.gradle:6]: Jcenter will become read only in May and completely disabled next year. I would not include it.

File: experimental/app/build.gradle [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+dependencies {
+
+    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+    implementation 'androidx.core:core-ktx:1.3.2'
+    implementation 'androidx.appcompat:appcompat:1.2.0'
+    implementation 'com.google.android.material:material:1.3.0'
+    implementation 'org.slf4j:slf4j-api:1.7.30'
+    implementation 'com.github.tony19:logback-android:2.0.0'
+    testImplementation 'junit:junit:4.+'
```
[the remark this answers is not in the document]
Comment (pcrepieux) [reply experimental/app/build.gradle:59]: 👍

File: experimental/app/src/main/java/io/devicefarmer/minicap/output/MinicapClientOutput.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        ByteBuffer.wrap(byteArray).apply {
+            order(ByteOrder.LITTLE_ENDIAN)
+            put(BANNER_VERSION.toByte())
+            put(BANNER_SIZE.toByte())
+            putInt(android.os.Process.myPid()) //PID
+            putInt(screenSize.width)//Width
+            putInt(screenSize.height)//Height
+            putInt(targetSize.width)//resized Width
+            putInt(targetSize.height)//resized height
+            put(0.toByte()) //orientation
```
[the remark this answers is not in the document]
Comment (pcrepieux) [reply experimental/app/src/main/java/io/devicefarmer/minicap/output/MinicapClientOutput.kt:55]: Or maybe should I simply have those sizes as parameters of sendBanner. They actually aren't used anywhere else at the moment.

File: experimental/app/src/main/java/io/devicefarmer/minicap/utils/DisplayManagerGlobal.kt (hunk 1 above)
Comment (pcrepieux) [reply experimental/app/src/main/java/io/devicefarmer/minicap/utils/DisplayManagerGlobal.kt:27]: 👍

File: experimental/build.gradle (hunk 2 above)
Comment (pcrepieux) [reply experimental/build.gradle:6]: 👍

**Review by koral-- — APPROVED** [review]

**Comment by varundtsfi** [comment]
I am not aware about the kotlin but I will use this approach.:+1:

**Pull request merged**